dataflow/src/main/java/org/checkerframework/dataflow/cfg/AbstractCFGVisualizer.java

```diff
@@ -0,0 +1,337 @@
+                                ccur.getId(),
+                                thenSuccessor.getId(),
+                                thenLabelContent + ccur.getThenFlowRule()));
+                if (!visited.contains(thenSuccessor)) {
+                    visited.add(thenSuccessor);
+                    workList.add(thenSuccessor);
+                }
+                Block elseSuccessor = ccur.getElseSuccessor();
+                sbDigraph.append(
+                        addEdge(
+                                ccur.getId(),
+                                elseSuccessor.getId(),
+                                elseLabelContent + ccur.getElseFlowRule()));
+                if (!visited.contains(elseSuccessor)) {
+                    visited.add(elseSuccessor);
+                    workList.add(elseSuccessor);
+                }
+            } else {
+                assert cur instanceof SingleSuccessorBlock;
+                Block b = ((SingleSuccessorBlock) cur).getSuccessor();
+                if (b != null) {
+                    sbDigraph.append(
+                            addEdge(
+                                    cur.getId(),
+                                    b.getId(),
+                                    ((SingleSuccessorBlock) cur).getFlowRule().name()));
+                    if (!visited.contains(b)) {
+                        visited.add(b);
+                        workList.add(b);
+                    }
+                }
+            }
+
+            // exceptional edges
+            if (cur.getType() == Block.BlockType.EXCEPTION_BLOCK) {
+                ExceptionBlock ecur = (ExceptionBlock) cur;
+                for (Map.Entry<TypeMirror, Set<Block>> e :
+                        ecur.getExceptionalSuccessors().entrySet()) {
+                    Set<Block> blocks = e.getValue();
+                    TypeMirror cause = e.getKey();
+                    String exception = cause.toString();
+                    if (exception.startsWith("java.lang.")) {
+                        exception = exception.replace("java.lang.", "");
+                    }
+
+                    for (Block b : blocks) {
+                        sbDigraph.append(addEdge(cur.getId(), b.getId(), exception));
+                        if (!visited.contains(b)) {
+                            visited.add(b);
+                            workList.add(b);
+                        }
+                    }
+                }
+            }
+
+            cur = workList.poll();
+        }
+
+        sbDigraph.append(generateNodes(visited, cfg, analysis));
+
+        // footer
+        sbDigraph.append("}\n");
+
+        return sbDigraph.toString();
+    }
+
+    protected abstract String generateNodes(
+            Set<Block> visited, ControlFlowGraph cfg, @Nullable Analysis<A, S, T> analysis);
+
+    protected String addEdge(long sId, long eId, String labelContent) {
+        return "    " + sId + " -> " + eId + " [label=\"" + labelContent + "\"];\n";
```

Comment:
This is also dot format. Maybe just leave this method abstract. `labelContent` isn't the most descriptive name, when it is not used as a dot label.